### ISSUE TITLE
Add GMAO model support

### ISSUE THREAD

**Issue by jlmaurer** [dbekaert/RAiDER#36]

**Comment by dbekaert** [comment]
see item #35 as there is a lot of similarity between what is needed for MERRA-2.
e.g. likely leverage same OpenDAP mechanism (different base URL) and same loading mechanism (netcdf4) for the variables. In terms of the model levels see below PDF for information

Relevant URLs:
- background information: https://gmao.gsfc.nasa.gov/GMAO_products/documents/GEOS_5_FP_File_Specification_ON4v1_2.pdf
- website for data-access (see left openDAP data): https://fluid.nccs.nasa.gov/weather/
 
Data to be used:
- **preferred:** inst3_3d_asm_Nv: 3Hour outputs are 0.3° lateral sampling and 72 vertical model levels. The information on the model levels is in the above PDF
- inst3_3d_asm_Np: 3Hour outputs at 0.3° Lateral sampling and 42 pressure levels

**Comment by leiyangleon** [comment]
@dbekaert Using the preferred model level [URL](https://opendap.nccs.nasa.gov/dods/GEOS-5/fp/0.25_deg/assim/inst3_3d_asm_Nv
), I am able to get the dataset returned but not able to retrieve the specific data values. The line I used is
`ds['t'].array[10:14,10:14,10:14,10:14].data`
But the error says **_IndexError: only integers, slices (`:`), ellipsis (`...`), numpy.newaxis (`None`) and integer or boolean arrays are valid indices_**
I tried the example on the Pydap Python [Wikipage](https://www.pydap.org/en/latest/client.html) and it worked fine.


Also, as mentioned [here](https://opendap.nccs.nasa.gov/dods/GEOS-5/fp/0.25_deg/assim/inst3_3d_asm_Nv.info), the temporal resolution is 3 hrs. But you wanted to use 7.5 hrs for the analysis data. So does it require a different URL? Or we can code that 7.5 hrs in the reader and still use the same URL?

**Comment by dbekaert** [comment]
@leiyangleon the model sampling is indeed every 3 hours. The 7.5hr latency means a user cannot request any data within 7.5 hour from real-time as it does not yet exist. We should capture that an error out to the user for such requests.

Can you provide more information on what you are doing in terms of getting the data using OpenDAP, for recreation purposes? I assume you might already slicing through the data? E.g. What happens if you ask for the complete array of 't' instead?

**Comment by leiyangleon** [comment]
@dbekaert Thanks for the clarification on the latency vs temporal resolution. 

Below are the lines I used to call OpenDAP for GMAO model:

```
import pydap.client
import pydap.cas.urs
url = 'https://opendap.nccs.nasa.gov/dods/GEOS-5/fp/0.25_deg/assim/inst3_3d_asm_Nv'
session = pydap.cas.urs.setup_session('username', 'password', check_url=url)
ds = pydap.client.open_url(url, session=session)
ds['t'].array[10:14,10:14,10:14,10:14].data
```

No matter whether I ask for the whole array or a sliced version, it cannot retrieve the specific values. For example, if I use `A=ds['t'].array.data` and then retrieve the values using `A[10,10,10,10]`, the same error popped up.

**Comment by jlmaurer** [comment]
@leiyangleon you may need to have the right extension? I.e. https://opendap.nccs.nasa.gov/dods/GEOS-5/fp/0.25_deg/assim/inst3_3d_asm_Nv.das --> note the .das extension.

**Comment by dbekaert** [comment]
@leiyangleon.

Similar to the merra model i believe you need to add the GSFC DAAC  application to your earth data profile. 
Also per email conversation with GMAO, see here for additional documentation: https://colab.research.google.com/drive/1f_ss1Oa3VzgAOd_p8sgekdnLVE5NW6s5

**Comment by leiyangleon** [comment]
@dbekaert @jlmaurer After reading the [documentation](https://colab.research.google.com/drive/1f_ss1Oa3VzgAOd_p8sgekdnLVE5NW6s5), I have finally figured out the problem. In the documentation, it says PyDAP is not working due to a known bug. Since we do want to use PyDAP, I then went to the two posts about the bug, which is exactly the same error as I reported. Here they are: [post1](https://github.com/pydap/pydap/issues/121) and [post2](https://github.com/pydata/xarray/issues/1860). Although there is no workable solution to solve this, I did notice that people reported this error associated with PyDAP 3.2.2 while PyDAP 3.2.1 worked fine and they are currently waiting for the OpenDAP team to fix that and release a new version 3.2.3. 

After checking my install, I do have PyDAP 3.2.2. So I downgraded my PyDAP to 3.2.1 which also downgraded my Python from 3.8 to 3.7.6. Then I finally managed to import the data array from the above GMAO [link](https://opendap.nccs.nasa.gov/dods/GEOS-5/fp/0.25_deg/assim/inst3_3d_asm_Nv) 👍 (@jlmaurer it is fine without the das extension; @dbekaert unlike MERRA-2, it is fine without adding the application `NASA GESDISC DATA ARCHIVE` to my earth data profile). I guess downgrading Python to 3.7.6 won't impact RAiDER's performance at this stage (correct me if I am wrong). So we have this temporary solution of appending 3.2.1 to PyDAP in the environment.yml file. In the long run, once the OpenDAP team fixes that bug in a new release 3.2.3, we can always update the PyDAP version then.

**Comment by dbekaert** [comment]
@leiyangleon did you add the application to your earthdata profile for merra-1 before testing or after testing the download of GMAO?

**Comment by leiyangleon** [comment]
@dbekaert yes will update both the linked file and the readme.

I tested the application for each of MERRA-2 and GMAO independently by adding/removing the Earthdata application. MERRA-2 is sensitive to the application while GMAO is not.

**Comment by jlmaurer** [comment]
@leiyangleon that is great! I believe we can specify to exclude a specific person in the .yml file.

**Comment by dbekaert** [comment]
> @dbekaert yes will update both the linked file and the readme.
> 
> I tested the application for each of MERRA-2 and GMAO independently by adding/removing the Earthdata application. MERRA-2 is sensitive to the application while GMAO is not.

Ah great, i was not aware that is possible!  Thanks for confirming.

**Comment by leiyangleon** [comment]
@jlmaurer @dbekaert I have made the GMAO model to run through (see [gmao.py](https://github.com/leiyangleon/RAiDER/blob/dev/tools/RAiDER/models/gmao.py)). However, there are a few things that I wanted to bring to your attention. 

1. the boundary elements of the 3-D cubes may contain nans after using "interpolate_along_axis". See the [codes](https://github.com/leiyangleon/RAiDER/blob/634e83712ef2ec65e3cae8ad089a45bddd8c1e3e/tools/RAiDER/models/gmao.py#L148-L152) I used. Although it does not seem to issue errors and let the program quit, I added a few [lines](https://github.com/leiyangleon/RAiDER/blob/634e83712ef2ec65e3cae8ad089a45bddd8c1e3e/tools/RAiDER/models/gmao.py#L154-L173) to fill the missing values based on linear (can be higher order as well) interpolation along vertical axis. I am not sure how you treat missing values later in the code, but my way may be useful to fix them at the beginning.

2. the other problem was related to what @jlmaurer suggested on using 1-D vectors of self._xs, self._ys and self._zs. Using these (by uncommenting the [lines](https://github.com/leiyangleon/RAiDER/blob/634e83712ef2ec65e3cae8ad089a45bddd8c1e3e/tools/RAiDER/models/gmao.py#L183-L185) and commenting out the [lines](https://github.com/leiyangleon/RAiDER/blob/634e83712ef2ec65e3cae8ad089a45bddd8c1e3e/tools/RAiDER/models/gmao.py#L186-L188)), we got the following error message:

`Traceback (most recent call last):
  File "/opt/anaconda3/envs/RAiDER/bin/raiderDelay.py", line 13, in <module>
    parseCMD()
  File "/opt/anaconda3/envs/RAiDER/lib/python3.7/site-packages/RAiDER/runProgram.py", line 154, in parseCMD
    outformat, t, out, download_only, verbose, wfn, hfn)
  File "/opt/anaconda3/envs/RAiDER/lib/python3.7/site-packages/RAiDER/delay.py", line 159, in tropo_delay
    lons=lons, los=los, zref = zref, time=time, verbose=verbose, download_only=download_only)
  File "/opt/anaconda3/envs/RAiDER/lib/python3.7/site-packages/RAiDER/processWM.py", line 80, in prepareWeatherModel
    weather_model.load(f, outLats = lats, outLons = lons, los = los, zref = zref)
  File "/opt/anaconda3/envs/RAiDER/lib/python3.7/site-packages/RAiDER/models/weatherModel.py", line 134, in load
    self._uniform_in_z(_zlevels=_zlevels)
  File "/opt/anaconda3/envs/RAiDER/lib/python3.7/site-packages/RAiDER/models/weatherModel.py", line 563, in _uniform_in_z
    _zlevels = np.nanmean(self._zs, axis=(0,1))
  File "<__array_function__ internals>", line 6, in nanmean
  File "/opt/anaconda3/envs/RAiDER/lib/python3.7/site-packages/numpy/lib/nanfunctions.py", line 950, in nanmean
    cnt = np.sum(~mask, axis=axis, dtype=np.intp, keepdims=keepdims)
  File "<__array_function__ internals>", line 6, in sum
  File "/opt/anaconda3/envs/RAiDER/lib/python3.7/site-packages/numpy/core/fromnumeric.py", line 2229, in sum
    initial=initial, where=where)
  File "/opt/anaconda3/envs/RAiDER/lib/python3.7/site-packages/numpy/core/fromnumeric.py", line 90, in _wrapreduction
    return ufunc.reduce(obj, axis, dtype, out, **passkwargs)
numpy.AxisError: axis 1 is out of bounds for array of dimension 1`

So I decided to use 3-D cubes of each variable, which was also used in other weather model readers, e.g. ERA5, ERA5T, HRRR. The code in its current version runs through without problems.

**Comment by dbekaert** [comment]
That is great!

> * the boundary elements of the 3-D cubes may contain nans after using "interpolate_along_axis". See the [codes](https://github.com/leiyangleon/RAiDER/blob/634e83712ef2ec65e3cae8ad089a45bddd8c1e3e/tools/RAiDER/models/gmao.py#L148-L152) I used. Although it does not seem to issue errors and let the program quit, I added a few [lines](https://github.com/leiyangleon/RAiDER/blob/634e83712ef2ec65e3cae8ad089a45bddd8c1e3e/tools/RAiDER/models/gmao.py#L154-L173) to fill the missing values based on linear (can be higher order as well) interpolation along vertical axis. I am not sure how you treat missing values later in the code, but my way may be useful to fix them at the beginning.
> *
@leiyangleon Would this be related problem be solved if you download more weather model notes than the user is requesting to process data upon? The amount of data to ask extra will depend on the weather model sampling. e.g HRRR is 3 km sampling GMAO is like 12 km?; so we could add a delta to the bbox to make sure we always download more data from the model than we need for the grid.



> * So I decided to use 3-D cubes of each variable, which was also used in other weather model readers, e.g. ERA5, ERA5T, HRRR. The code in its current version runs through without problems.
> * 
@leiyangleon this sounds good. Eventually: @jlmaurer a change would need to be the same for all models.

**Comment by leiyangleon** [comment]
@dbekaert I coded the [horizontal sampling](https://github.com/leiyangleon/RAiDER/blob/634e83712ef2ec65e3cae8ad089a45bddd8c1e3e/tools/RAiDER/models/gmao.py#L35-L39) in the reader and the bbox also have a delta (e.g. 0.5 deg) by default. So I think we did download more data than we need. The missing data actually relate to the interpolation along vertical axis not horizontal sampling or bbox, from what I saw. For example, there 72 model levels for GMAO where first 4 and last 2 levels have missing data located here and there.

**Comment by dbekaert** [comment]
@leiyangleon, generally the models do not provide observations under topography. Your approach is correct to extrapolate. 

I believe we also do this for the other models. @jlmaurer can you confirm? 
In terms of modularity would be good to leverage the same function for all models perhaps such we have 1 unit test for it.

**Comment by jlmaurer** [comment]
@dbekaert and @leiyangleon that is correct, we currently interpolate nan values vertically, but in the base weather model "load" method, so that should be taken care of automatically. @leiyangleon if you are getting nans after the weather model gets written to the HDF5 file that is a bug and should be fixed.

**Comment by leiyangleon** [comment]
@dbekaert @jlmaurer Without using my lines of filling missing values, I checked the variables that got saved in the h5 files, which don't include any missing values. So as @jlmaurer pointed out, yes, the base weather model "load" method indeed works to fill those missing values. Anyway, I have commented out my lines of filling missing values for now.

**Comment by dbekaert** [comment]
@leiyangleon sounds good feel and ready to be merged at your convenience.
I had a look at the documentation of the model too.
Suggest to remove  your <key> value as that is personal from the cdsa api instructions.

@Askaholic might argue against hashed out statements in code.

**Comment by Askaholic** [comment]
Yes, any commented out code should be deleted before merging. Either you need it, or you delete it :)

**Comment by dbekaert** [comment]
This has been closed out through #71
